### PR TITLE
 feat: more efficient detection of advisory formats

### LINKED ISSUE #257: Be more efficient with auto-detection of document format.s
There's probably performance opportunities in here, e.g. use `RawValue` for everything in the doc except the field[s] required to recognize the format. That way, we'd save a crapton of memory allocation and we'd know exactly what format the errors applied to. Also, we should probably order them by most common expected.

We can do that later, obviously.

_Originally posted by @jcrossley3 in https://github.com/trustification/trustify/pull/256#discussion_r1595822037_

### PR DESCRIPTION
**Pull request by jcrossley3**
Fixes #257

We only parse the documents enough to detect the fields identifying the format. We no longer parse a doc just to determine its format, so errors due to invalid data will be tossed downstream.

### CODE REVIEW

**Comment by ctron** [comment]
`jsn` looks cool!

**Review by bobmcwhirter — APPROVED** [review]

File: modules/ingestor/src/service/format.rs [hunk 1]
```diff
@@ -52,36 +51,47 @@ impl<'g> Format {
...
+            key("document").and(key("csaf_version")).and(depth(2)),
+            bytes,
+        ) {
+            Ok(Format::CSAF { checksum })
+        } else if masked(depth(1).and(key("dataType")), bytes) {
+            Ok(Format::CVE { checksum })
+        } else if masked(depth(1).and(key("id")), bytes) {
+            Ok(Format::OSV { checksum })
+        } else {
+            Err(Error::UnsupportedFormat("No recognized fields".into()))
```
[the remark this answers is not in the document]
Comment (bobmcwhirter) [reply modules/ingestor/src/service/format.rs:64]: Yes

[the remark this answers is not in the document]
Comment (jcrossley3) [reply modules/ingestor/src/service/format.rs:64]: I know X Y or Z. I don't know V1 and V2. I'm guessing you don't either. So what _exactly_ do you want the error message to say?

**Comment by bobmcwhirter** [comment]
I guess... discover what versions the auto-detect can detect (maybe we need more detection for CycloneDX 1.3 vs 1.4 for instance) and then wordsmith up something.

File: modules/ingestor/src/service/format.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcrossley3) [reply modules/ingestor/src/service/format.rs:64]: Am I correctly interpreting your answer by assuming we currently support *ALL* versions of CSAF, OSV, and CVE? Are there any we explicitly don't support?

**Comment by bobmcwhirter** [comment]
No I'm certain it's only a subset currently. But we should do all. And detect versions appropriately if required.

**Review by bobmcwhirter — APPROVED** [review]

**Pull request merged**